llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp

```diff
@@ -1149,15 +1143,26 @@ void RISCVInsertVSETVLI::insertVSETVLI(MachineBasicBlock &MBB,
                 .addImm(Info.encodeVTYPE());
   if (LIS) {
     LIS->InsertMachineInstrInMaps(*MI);
-    // Normally the AVL's live range will already extend past the inserted
-    // vsetvli because the pseudos below will already use the AVL. But this
-    // isn't always the case, e.g. PseudoVMV_X_S doesn't have an AVL operand or
-    // we've taken the AVL from the VL output of another vsetvli.
     LiveInterval &LI = LIS->getInterval(AVLReg);
     SlotIndex SI = LIS->getInstructionIndex(*MI).getRegSlot();
-    assert((LI.liveAt(SI) && LI.getVNInfoAt(SI) == Info.getAVLVNInfo()) ||
-           (!LI.liveAt(SI) && LI.containsOneValue()));
-    LIS->extendToIndices(LI, SI);
+    // If the AVL value isn't live at MI, do a quick check to see if it's easily
+    // extendable. Otherwise, we need to copy it.
+    if (LI.getVNInfoBefore(SI) != Info.getAVLVNInfo()) {
+      if (!LI.liveAt(SI) && LI.containsOneValue())
+        LIS->extendToIndices(LI, SI);
+      else {
+        Register AVLCopyReg =
+            MRI->createVirtualRegister(&RISCV::GPRNoX0RegClass);
+        MachineBasicBlock::iterator AVLDef =
```

Comment:
Do you know how a PHIDef is handled here?  Is the slot index simply mapped to the first instruction in the block?  Or something else?  Either way, we may need special handling for that case here. 

For correctness, I'd appreciate if you did a local patch which removes the previous extend check and ensures everything works.  The fallback path should be generic.

Reply:
A PHI def's SlotIndex will be at the entry to the block, so it will insert a copy at the entry, which I believe is what we want.

>  I'd appreciate if you did a local patch which removes the previous extend check and ensures everything works

If I'm understanding you correctly, the previous extend check is removed in this patch already

Reply:
> A PHI def's SlotIndex will be at the entry to the block, so it will insert a copy at the entry, which I believe is what we want.

Are we guaranteed that the Slot_Block entry maps to an MachineInstr?  Getting a MachineInstr can return nullptr, and I'm not clear about the exact circumstances.

> > I'd appreciate if you did a local patch which removes the previous extend check and ensures everything works
> 
> If I'm understanding you correctly, the previous extend check is removed in this patch already

You have:
```
      if (!LI.liveAt(SI) && LI.containsOneValue())
        LIS->extendToIndices(LI, SI);
```

I'm suggesting run "ninja check" (and other tests of your choice) with:

```
      if (false && !LI.liveAt(SI) && LI.containsOneValue())
        LIS->extendToIndices(LI, SI);
```

i.e. Force *everything* done the COPY path.  This will hurt code quality, but should never violate invariants.  (This is only a local to you test to make sure it works.)

Reply:
> Are we guaranteed that the Slot_Block entry maps to an MachineInstr? Getting a MachineInstr can return nullptr, and I'm not clear about the exact circumstances.

Thanks for catching that, turns out`getInstructionFromIndex` does return null for that. I've added explicit handling for this to insert it with the other PHIs. We also didn't have any test cases that hit this with a PHI so I've added one. 

Sorry I see what you mean about removing the extend check now, I got it confused with the check on line 953. I can confirm that does work and is generic.